src/java.scripting/share/classes/javax/script/ScriptEngineManager.java

```diff
@@ -311,41 +209,50 @@ public ScriptEngine getEngineByExtension(String extension) {
      * @throws NullPointerException if mimeType is null.
      */
     public ScriptEngine getEngineByMimeType(String mimeType) {
-        if (mimeType == null) throw new NullPointerException();
-        //look for registered types first
-        Object obj;
-        if (null != (obj = mimeTypeAssociations.get(mimeType))) {
-            ScriptEngineFactory spi = (ScriptEngineFactory)obj;
-            try {
-                ScriptEngine engine = spi.getScriptEngine();
-                engine.setBindings(getBindings(), ScriptContext.GLOBAL_SCOPE);
-                return engine;
-            } catch (Exception exp) {
-                if (DEBUG) exp.printStackTrace();
-            }
-        }
+        return getEngineBy(mimeType, mimeTypeAssociations, ScriptEngineFactory::getMimeTypes);
+    }
 
-        for (ScriptEngineFactory spi : engineSpis) {
-            List<String> types = null;
-            try {
-                types = spi.getMimeTypes();
-            } catch (Exception exp) {
-                if (DEBUG) exp.printStackTrace();
-            }
-            if (types == null) continue;
-            for (String type : types) {
-                if (mimeType.equals(type)) {
-                    try {
-                        ScriptEngine engine = spi.getScriptEngine();
-                        engine.setBindings(getBindings(), ScriptContext.GLOBAL_SCOPE);
-                        return engine;
-                    } catch (Exception exp) {
-                        if (DEBUG) exp.printStackTrace();
-                    }
+    private ScriptEngine getEngineBy(String selector, Map<String, ScriptEngineFactory> associations, Function<ScriptEngineFactory, List<String>> valuesFn) {
+        Objects.requireNonNull(selector);
+        Stream<ScriptEngineFactory> spis = Stream.concat(
+            //look for registered types first
+            Stream.ofNullable(associations.get(selector)),
+
+            engineSpis.stream().filter(spi -> {
+                try {
+                    List<String> matches = valuesFn.apply(spi);
+                    return matches != null && matches.contains(selector);
```

Comment:
Note this under anomalous conditions, calling `List.contains` here can result in somewhat different behavior than [what was here before](https://github.com/openjdk/jdk/blob/master/src/java.scripting/share/classes/javax/script/ScriptEngineManager.java#L238). Notably:
* if `spi` returns a list that contains the selector value _at least twice_, and
* it throws an exception while creating the engine for the first time, then
* the previous implementation will invoke `getScriptEngine` for the second time too.

My modified implementation will ignore multiple occurrences of the selector value in the returned list. I believe that behavior is correct, especially since the list of values for engine names, extensions, and mime types is expected to contain unique entries (too bad the interface didn't define these methods to return a `Set<String>`…)